### PR TITLE
Upgrade to 4.22 with backwards comparability for 4.20

### PR DESCRIPTION
**Pull request by Vatyx**
#### Description
The upgrade to 4.22 mostly consisted of changing function calls to use the correct parameters. If functionality is the same with this PR for 4.20. We can have this go in while I wrap up work regarding the Engine and Documentation.

Engine PR (just for looking at comparisons): https://github.com/improbableio/UnrealEngine/pull/128

#### Tests
Tested it with shooter game on 4.22.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp [hunk 1]
```diff
@@ -335,7 +335,7 @@ bool CreateStartupActor(Worker_SnapshotOutputStream* OutputStream, AActor* Actor
 		});
 	}
 
-	USpatialActorChannel* Channel = Cast<USpatialActorChannel>(NetConnection->CreateChannel(CHTYPE_Actor, 1));
+	USpatialActorChannel* Channel = nullptr; //Cast<USpatialActorChannel>(NetConnection->CreateChannelByName(NAME_Actor, EChannelCreateFlags::OpenedLocally));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp]: er......

File: SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp [hunk 2]
```diff
@@ -195,7 +195,7 @@ bool CreatePlaceholders(Worker_SnapshotOutputStream* OutputStream)
 void SetupStartupActorCreation(USpatialNetDriver*& NetDriver, USpatialNetConnection*& NetConnection, USpatialPackageMapClient*& PackageMap, USpatialClassInfoManager*& ClassInfoManager, UWorld* World)
 {
 	NetDriver = NewObject<USpatialNetDriver>();
-	NetDriver->ChannelClasses[CHTYPE_Actor] = USpatialActorChannel::StaticClass();
+	//NetDriver->ChannelClasses[CHTYPE_Actor] = USpatialActorChannel::StaticClass();
```
[the remark this answers is not in the document]
Comment (Vatyx) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp]: These are the left over functions from the Startup Actor snapshot generation so we can have them for reference. I should probably fix the comment.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp (hunk 1 above)
Comment (Vatyx) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp]: Right above the function there is a comment saying these functions are not in use

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 3]
```diff
@@ -1422,23 +1444,36 @@ void USpatialReceiver::ResolveObjectReferences(FRepLayout& RepLayout, UObject* R
...
+#else
+		int32 StoredDataOffset = Cmd->ShadowOffset;
+#endif
+
 		if (ObjectReferences.Array)
 		{
 			check(Property->IsA<UArrayProperty>());
 
-			Property->CopySingleValue(StoredData + AbsOffset, Data + AbsOffset);
+			if (!bIsHandover)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: The logic for this functionality is quite different now dependent on if it's handover or not, perhaps it's better to be more explicit up front and pull it into two sections of functionality.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 4]
```diff
@@ -124,19 +140,19 @@ bool USpatialActorChannel::IsSingletonEntity()
...
+#else
+bool USpatialActorChannel::CleanUp(const bool bForDestroy, EChannelCloseReason CloseReason)
+#endif
 {
 #if WITH_EDITOR
 	if (NetDriver != nullptr)
 	{
 		const bool bDeleteDynamicEntities = GetDefault<ULevelEditorPlaySettings>()->GetDeleteDynamicEntities();
 
-		UWorld* World = NetDriver->GetWorld();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Why was this removed?

Comment (Vatyx) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: World no longer exists at that point in time when shutting down.

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: What about in 4.20?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentReader.cpp [hunk 5]
```diff
@@ -133,7 +137,7 @@ void ComponentReader::ApplySchemaObject(Schema_Object* ComponentObject, UObject*
 
 					if (NewUnresolvedRefs.Num() > 0)
 					{
-						RootObjectReferencesMap.Add(SwappedCmd.Offset, FObjectReferences(ValueData, CountBits, NewUnresolvedRefs, Cmd.ParentIndex, ArrayProperty, /* bFastArrayProp */ true));
+						RootObjectReferencesMap.Add(SwappedCmd.Offset, FObjectReferences(ValueData, CountBits, NewUnresolvedRefs, BaseHandleToCmdIndex[FieldId - 1].CmdIndex, Cmd.ParentIndex, ArrayProperty, /* bFastArrayProp */ true));
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentReader.cpp]: Make `BaseHandleToCmdIndex[FieldId - 1].CmdIndex` a variable just before `const FRepLayoutCmd& Cmd = ...`

Comment (Vatyx) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentReader.cpp]: Sure thing.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentReader.cpp [hunk 6]
```diff
@@ -385,15 +394,15 @@ void ComponentReader::ApplyArray(Schema_Object* Object, Schema_FieldId FieldId,
 	for (int i = 0; i < Count; i++)
 	{
 		int32 ElementOffset = i * Property->Inner->ElementSize;
-		ApplyProperty(Object, FieldId, *ArrayObjectReferences, i, Property->Inner, ArrayHelper.GetRawPtr(i), ElementOffset, ParentIndex);
+		ApplyProperty(Object, FieldId, *ArrayObjectReferences, i, Property->Inner, ArrayHelper.GetRawPtr(i), ElementOffset, CmdIndex, ParentIndex);
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentReader.cpp]: This looks wrong. Have you tested with arrays of `UObject*`? It looks like when resolving an element of the array, it will use the shadow offset of the array within the replicating object, not the offset of the element within the array.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 4 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Was there any follow up on this. This was added in https://github.com/spatialos/UnrealGDK/commit/9b24edd369aac989ce6ac9337b5fb35c76b33b0c which still looks valid.

Comment (Vatyx) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Talked to @improbable-valentyn about this. Seems that this is fine considering since it only happens with the editor.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp [hunk 7]
```diff
@@ -192,235 +192,235 @@ bool CreatePlaceholders(Worker_SnapshotOutputStream* OutputStream)
...
-	NetConnection->Driver = NetDriver;
-	NetConnection->State = USOCK_Closed;
-
-	PackageMap = NewObject<USpatialPackageMapClient>();
-	PackageMap->Initialize(NetConnection, NetDriver->GuidCache);
-
-	NetConnection->PackageMap = PackageMap;
-	NetDriver->PackageMap = PackageMap;
-}
+//void SetupStartupActorCreation(USpatialNetDriver*& NetDriver, USpatialNetConnection*& NetConnection, USpatialPackageMapClient*& PackageMap, USpatialClassInfoManager*& ClassInfoManager, UWorld* World)
```
Comment (mattyoung-improbable) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp]: What's going on with all this commented out?

Comment (Vatyx) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp]: this is for reference since it was hard to do this initially

Comment (Vatyx) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp]: Though we can delete it I guess. Mike also wants to delete it.

Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp]: +1 to deleting it
